### ISSUE TITLE
add Android support

### ISSUE THREAD

**Issue by slimsag** [azul3d/oldissues#9]
With Go 1.4 supporting Android and targeting game development generally, we should begin working towards supporting Android.

**Comment by crawshaw** [comment]
I stumbled across this issue while searching for information about GL batching, I was hoping you could enlighten me.

I understand why this would be a significant issue in OpenGL 1.x, which is driven by a large number of function calls, but it seems to me that the most efficient form of batching in the shader world is actually to merge draw calls. Do you still end up with enough function calls that the cgo boundary is expensive enough to notice on a mobile device?

If so, I would be open to exploring a batching implementation inside golang.org/x/mobile/gl. I don't think any API changes are necessary, a batch queue could be built underneath what's there. But before doing so, I'd need an example program that would benefit from it, so I could measure the performance improvement.

**Comment by dskinner** [comment]
@crawshaw until @slimsag gets back to you, he put together a benchmark specific to batching that you can look over here that he'll likely elaborate on for you: https://github.com/slimsag/cgo-batching

and related discussion: https://github.com/azul3d/issues/issues/17

**Comment by slimsag** [comment]
@crawshaw 

Thank you for commenting here -- the situation has changed since I last posted here back in August.

> I stumbled across this issue while searching for information about GL batching, I was hoping you could enlighten me.

For nearly 90% of the work you are doing with OpenGL ES 2 era hardware, grouping together draw calls will benefit you the most.

> I understand why this would be a significant issue in OpenGL 1.x, which is driven by a large number of function calls, but it seems to me that the most efficient form of batching in the shader world is actually to merge draw calls. Do you still end up with enough function calls that the cgo boundary is expensive enough to notice on a mobile device?

As @dskinner did put, call batching is _not fictional_ in the sense that it does have measurable performance gains -- but as of today there isn't an application that suffers from this. For this reason we've already switched to Glow's OpenGL ES 2 bindings (github.com/go-gl/glow).

I think your bindings are more clean and albeit more idiomatic Go -- but we also use Glow for (desktop) OpenGL 2 bindings and it would be easier to add CGO call batching to one set of bindings rather than two individual ones, _should the need arise_.

> If so, I would be open to exploring a batching implementation inside golang.org/x/mobile/gl. I don't think any API changes are necessary, a batch queue could be built underneath what's there. But before doing so, I'd need an example program that would benefit from it, so I could measure the performance improvement.

I think if we did find a situation in which an application is severely hindered by this, it would be worth looking into improving this in CGO itself. It is not strictly an OpenGL problem but rather a CGO overhead cost, due to the fact that Go and C use different (and incompatible) calling conventions. Go's calling convention probably cannot be changed to match the C one, but perhaps CGO could feature implicit or explicit call batching.
